packages/data/src/components/use-select/index.js

```diff
@@ -40,6 +33,128 @@ const renderQueue = createQueue();
+		// in the interval between the `getValue` call during render and creating
+		// the subscription, which is slightly delayed. We need to ensure that this
+		// second `getValue` call will compute a fresh value.
+		lastMapResultValid = false;
+
+		const onStoreChange = () => {
+			// Invalidate the value on store update, so that a fresh value is computed.
+			lastMapResultValid = false;
+			listener();
+		};
+
+		const onChange = () => {
+			if ( lastIsAsync ) {
+				renderQueue.add( queueContext, onStoreChange );
+			} else {
+				onStoreChange();
+			}
+		};
+
+		const unsubs = stores.map( ( storeName ) => {
+			return registry.subscribe( onChange, storeName );
+		} );
+
+		return () => {
+			// The return value of the subscribe function could be undefined if the store is a custom generic store.
+			for ( const unsub of unsubs ) {
+				unsub?.();
+			}
+			// Cancel existing store updates that were already scheduled.
+			renderQueue.cancel( queueContext );
+		};
+	};
+
+	return ( mapSelect, resubscribe, isAsync ) => {
+		const selectValue = () => mapSelect( select, registry );
+
+		function updateValue( selectFromStore ) {
+			// If the last value is valid, and the `mapSelect` callback hasn't changed,
+			// then we can safely return the cached value. The value can change only on
+			// store update, and in that case value will be invalidated by the listener.
+			if ( lastMapResultValid && mapSelect === lastMapSelect ) {
+				return lastMapResult;
+			}
+
+			const mapResult = selectFromStore();
+
+			// If the new value is shallow-equal to the old one, keep the old one so
+			// that we don't trigger unwanted updates that do a `===` check.
+			if ( ! isShallowEqual( lastMapResult, mapResult ) ) {
+				lastMapResult = mapResult;
+			}
+			lastMapResultValid = true;
+		}
+
+		function getValue() {
+			// Update the value in case it's been invalidated or `mapSelect` has changed.
+			updateValue( selectValue );
+			return lastMapResult;
+		}
+
+		// When transitioning from async to sync mode, cancel existing store updates
+		// that have been scheduled, and invalidate the value so that it's freshly
+		// computed. It might have been changed by the update we just cancelled.
+		if ( lastIsAsync && ! isAsync ) {
+			lastMapResultValid = false;
+			renderQueue.cancel( queueContext );
+		}
+
+		// Either initialize the `subscribe` function, or create a new one if `mapSelect`
+		// changed and has dependencies.
+		// Usage without dependencies, `useSelect( ( s ) => { ... } )`, will subscribe
+		// only once, at mount, and won't resubscibe even if `mapSelect` changes.
+		if ( ! subscribe || ( resubscribe && mapSelect !== lastMapSelect ) ) {
+			// Find out what stores the `mapSelect` callback is selecting from and
+			// use that list to create subscriptions to specific stores.
+			const listeningStores = { current: null };
+			updateValue( () =>
+				registry.__unstableMarkListeningStores(
```

Comment:
Since this is done only on initialization, it means we assume the subscribed "stores" don't change as a response to a store data change? Maybe I'm missing something but I'm not sure this assumption is always correct.

Reply:
Yes, we assume that a given instance of `mapSelect` function always selects the same stores. Store data change can only change the results of the selects, but not the actual calls. In other words, a `mapSelect` function like this is unreliable:
```js
( select ) => select( 'a' ).bOrC() ? select( 'b' ).get() : select( 'c' ).get()
```
because it will subscribe only to either `b` or `c` but never to both. Depending on what got selected on the first call.

But this is already an existing behavior in the old `useSelect`, I'm merely reproducing it here.

Reply:
> But this is already an existing behavior in the old useSelect, I'm merely reproducing it here.

Interesting, I wonder if we should fix that behavior (not necessarily in this PR)

Reply:
> Store data change can only change the results of the selects, but not the actual calls. In other words, a `mapSelect` function like this is unreliable

Could you elaborate a little bit more on this? I thought that this was always supported 😅 .

Reply:
We can try that, but I'd prefer another PR because at the moment I'm not sure if it's really feasible. When calling `updateValue` inside a store listener callback, we can't resubscribe at that moment -- it must happen in an effect, after `useSyncExternalStore` returns a new `subscribe` function. But we're not in a `useSyncExternalStore` call, we are in a store listener. The timings can get complicated.

Also there currently are no practical problems with the existing approach. Conditional selects either don't happen, or happen in a compatible way, like:
```js
useSelect( ( select ) => {
  const { bOrC } = select( 'a' );
  const { get: getB } = select( 'b' );
  const { get: getC } = select( 'c' );
  return bOrC() ? getB() : getC();
}, [] );
```
This pattern, select and destructure getters first and call them second, is very common.

Reply:
> I thought that this was always supported

In the original `useSelect`, if you look at the `useLayoutEffect` that does the subscribing, it runs only when the dependencies change (`depsChangedFlag`). With `useSelect` without dependencies, the deps default to `[]`, i.e., the hook subscribes only on mount, to the stores that were called during the first call to `mapSelect`. When `useSelect` is called with explicit dependencies, subscribed stores can change only when dependencies change. If `mapSelect` selects from different stores on every call, based on store state and not captured in dependencies, then re-subscription never happens.

This has been true for a very long time, maybe ever since granular subscriptions were introduced.

Reply:
Since it doesn't seem to change the behavior and I agree that it's probably very uncommon behavior, let's leave it as is for now.

Reply:
I see! FWIW, I think this might be an oversight by me back in #26724 though. We should resubscribe when `listeningStores` changes. I did an ad-hoc change locally and it passed all unit tests. I agree that we should do it in another PR though.